libs/openFrameworks/gl/ofCubeMap.cpp

```diff
@@ -0,0 +1,804 @@
+#include "ofImage.h"
+#include "ofConstants.h"
+#include "of3dUtils.h"
+#include "ofGLBaseTypes.h"
+#include "ofGLUtils.h"
+
+#include "glm/gtx/transform.hpp"
+#include "glm/gtc/quaternion.hpp"
+
+#include "ofGLProgrammableRenderer.h"
+#include "ofCubeMapShaders.h"
+
+#ifndef STRINGIFY
+#define STRINGIFY(x) #x
+#endif
+
+using std::weak_ptr;
+using std::vector;
+using std::shared_ptr;
+
+ofVboMesh ofCubeMap::sCubeMesh;
+ofShader ofCubeMap::shaderBrdfLUT;
+ofTexture ofCubeMap::sBrdfLutTex;
+
+//----------------------------------------
+vector<weak_ptr<ofCubeMap::Data> > & ofCubeMapsData(){
+	static vector<weak_ptr<ofCubeMap::Data> > * cubeMapsActive = new vector<weak_ptr<ofCubeMap::Data> >;
+	return *cubeMapsActive;
+}
+
+//--------------------------------------------------------------
+bool ofCubeMap::hasActiveCubeMap() {
+	for(size_t i=0;i< ofCubeMapsData().size();i++){
+		std::shared_ptr<ofCubeMap::Data> cubeMap = ofCubeMapsData()[i].lock();
+		if(cubeMap && cubeMap->isEnabled && cubeMap->index > -1 ){
+			return true;
+			break;
+		}
+	}
+	return false;
+}
+
+//--------------------------------------------------------------
+std::shared_ptr<ofCubeMap::Data> ofCubeMap::getActiveData() {
+	for(size_t i=0;i< ofCubeMapsData().size();i++){
+		std::shared_ptr<ofCubeMap::Data> cubeMap = ofCubeMapsData()[i].lock();
+		if(cubeMap && cubeMap->isEnabled && cubeMap->index > -1 ){
+			return cubeMap;
+		}
+	}
+	return std::shared_ptr<ofCubeMap::Data>();
+}
+
+//--------------------------------------------------------------
+void ofCubeMap::_checkSetup() {
+	if( data->index < 0 ) {
+		bool bFound = false;
+		// search for the first free block
+		for(size_t i=0; i<ofCubeMapsData().size(); i++) {
+			if(ofCubeMapsData()[i].expired()) {
+				data->index = i;
+				ofCubeMapsData()[i] = data;
+				bFound = true;
+				break;
+			}
+		}
+		if(!bFound && ofIsGLProgrammableRenderer()){
+			ofCubeMapsData().push_back(data);
+			data->index = ofCubeMapsData().size() - 1;
+			bFound = true;
+		}
+	}
+	// we should remove empty slots //
+}
+
+//----------------------------------------
+const ofTexture& ofCubeMap::getBrdfLutTexture() {
+	return sBrdfLutTex;
+}
+
+//----------------------------------------
+ofCubeMap::ofCubeMap() {
+	texFormat = GL_RGB;
+	data = std::make_shared<ofCubeMap::Data>();
+#ifdef GL_TEXTURE_CUBE_MAP_SEAMLESS
+	glEnable(GL_TEXTURE_CUBE_MAP_SEAMLESS);
+#endif
+	_checkSetup();
+	projectionMat = glm::perspective(glm::radians(90.0f), 1.0f, 0.1f, 10.0f );
+	
+}
+
+//----------------------------------------
+ofCubeMap::~ofCubeMap() {
+	if( data ) {
+		if( data->bPreFilteredMapAllocated ) {
+			data->bPreFilteredMapAllocated=false;
+			glDeleteTextures(1, &data->preFilteredMapId );
+		}
+		
+		if( data->bIrradianceAllocated ) {
+			data->bIrradianceAllocated = false;
+			glDeleteTextures(1, &data->irradianceMapId );
+		}
+		if( data->bCubeMapAllocated ) {
+			data->bCubeMapAllocated = false;
+			glDeleteTextures(1, &data->cubeMapId );
+		}
+	}
+}
+
```

Comment:
I wonder if we need a copy constructor for this? 
And/or texture reference counting ( retain ) similar to: 

https://github.com/openframeworks/openFrameworks/blob/master/libs/openFrameworks/gl/ofTexture.cpp#L187-L248

Reply:
Yeah would probably be beneficial